lib/postgres_v3_experimental.dart

```diff
@@ -0,0 +1,267 @@
+  }
+
+  Future<void> dispose();
+}
+
+class PgTypedParameter {
+  final PgDataType type;
+  final Object? value;
+
+  PgTypedParameter(this.type, this.value);
+}
+
+abstract class PgResult implements List<PgResultRow> {
+  int get affectedRows;
+  PgResultSchema get schema;
+
+  factory PgResult(
+          List<PgResultRow> rows, int affectedRows, PgResultSchema schema) =
+      _PgResult;
+}
+
+class _PgResult extends DelegatingList<PgResultRow> implements PgResult {
+  @override
+  final int affectedRows;
+
+  @override
+  final PgResultSchema schema;
+
+  final List<PgResultRow> rows;
+
+  _PgResult(this.rows, this.affectedRows, this.schema) : super(rows);
+
+  @override
+  String toString() {
+    return 'PgResult(schema = $schema, affectedRows = $affectedRows, rows = $rows)';
+  }
+}
+
+abstract class PgResultRow implements List<Object?> {
+  PgResultSchema get schema;
+}
+
+class PgResultSchema {
+  final List<PgResultColumn> columns;
+
+  PgResultSchema(this.columns);
+
+  @override
+  String toString() {
+    return 'PgResultSchema(${columns.join(', ')})';
+  }
+}
+
+class PgResultColumn {
+  final PgDataType type;
+  final String? tableName;
+  final int? tableOid;
+  final String? columnName;
+  final int? columnOid;
+
+  PgResultColumn({
+    required this.type,
+    this.tableName,
+    this.tableOid,
+    this.columnName,
+    this.columnOid,
+  });
+
+  @override
+  String toString() {
+    final buffer = StringBuffer('${type.name} ');
+    if (tableName != null && tableName != '') {
+      buffer
+        ..write(tableName)
+        ..write('.');
+    } else if (tableOid != null && tableOid != 0) {
+      buffer
+        ..write('@$tableOid')
+        ..write('.');
+    }
+
+    if (columnName != null && columnName != '') {
+      buffer.write(columnName);
+    } else if (columnOid != null && columnOid != 0) {
+      buffer.write('@$columnOid');
+    }
+
+    return buffer.toString();
+  }
+}
+
+abstract class PgChannels {
+  Stream<String> operator [](String channel);
+  Future<void> notify(String channel, [String? payload]);
+  Future<void> cancelAll();
+}
+
+abstract class PgNotification {}
+
+class PgEndpoint {
+  final String host;
+  final int port;
+  final String database;
+  final String? username;
+  final String? password;
+  final bool requireSsl;
+  final bool isUnixSocket;
+
+  /// An optional [StreamChannelTransformer] sitting behind the postgres client
+  /// as implemented in the `posgres` package and the database server.
+  ///
+  /// The stream channel transformer is able to view, alter, drop, or inject
+  /// messages in either direction. This powerful tool can be used to implement
+  /// additional or custom functionality, but should also be used with caution
+  /// as altering the message flow might break internal invariants of this
+  /// package.
+  ///
+  /// For an example, see `example/v3/transformer.dart`.
+  final StreamChannelTransformer<BaseMessage, BaseMessage>? transformer;
```

Comment:
I think this should not live on the endpoint, which should be entirely serializable (e.g. read from configuration). Rather, it could live at the connection initializer method?

Reply:
I have moved it to the `PgSessionSettings` class, which is not serializable to begin with (due to the `badCertificate` callback).

I think it's more helpful to have it on `PgSessionSettings` than the connect methods since the class can be used for pools and single connections, it requires less duplication for application developers.